### ISSUE TITLE
Consistent naming of resources across operators

### ISSUE THREAD

**Issue by lfrancke** [stackabletech/issues#10]
Currently most (or all?) of the objects we create have "meaningful" names (i.e. they include things like the node name or role group). This is useful for debugging purposes but doesn't work in practice because the resulting names could be too long for Kubernetes.

Unfortunately, some parts of our code might rely on these names (e.g. fetch "this" configmap) while they should rely on labels or other forms of identification.

We should have a single generic way of handling name generation and we need to look through all our code to find places where we rely on the name of objects.

Kubernetes has a method of automatically generating unique names using the `generateName` field. We can still use meaningful prefixes.

**Comment by soenkeliebau** [comment]
I agree! The issue remains, even when using generateName though, as the prefix can still differ and sow confusion.
Would it make sense to have functions to generate the names in the framework and agree to exclusively use those?

**Comment by lfrancke** [comment]
Can you explain in more detail what you mean and what you propose?

My idea was just to use the "name" of the service for the pods plus a dash and then set generateName to this prefix.
So a ZooKeeperCluster called "production" would create prefixes "production-" which would lead to pod names like "production-hjlbf"

**Comment by soenkeliebau** [comment]
But wouldn't what you propose lead to confusion if there was an HDFS cluster also called production?

Overall I totally agree, my proposal was simply to extract the logic you described into a function in the framework to ensure that everybody uses the same logic.

**Comment by lfrancke** [comment]
At the Pod level it would be harder to distinguish, yes.
I've followed what Kubernetes is doing itself with Deployments/ReplicaSets etc.
We could include a short name (`zk-foo-`). Not sure if there are any drawbacks

**Comment by soenkeliebau** [comment]
I think the main drawback will usually be legibility/useability. Ideally I'd like to be able to have names that are sufficiently unique that someone logging onto the machine or looking at the list of pods can tell at a glance which pod/configmap/... belongs to which service and provides which product.
The main thing that I'd like to avoid is that a lookup in Kubernetes is necessary for any work on the actual server, for example to find out which pod provides Kafka and how the service is called.

The issue I can see with short names is that people can create confusing setups out of a lack of knowledge what this short name may influence, at creation time someone needs to specify a short name, doesn't really know what for and picks something, which later on collides with the dev system because this was forgotten at the time. Basically anything "user decided" I'd consider a danger, but anything shortened by us automatically may become unreadable..

TL/DR: I have opinions but no solution.

**Comment by lfrancke** [comment]
Sorry. What I meant was different.

Someone creates a ZooKeeper with the name `production`
The ZooKeeper operator uses its own short name (could be made configurable too) `zk` so all Pods end up being called `zk-production-<random>`

Thinking about it: Having a configurable `pod_prefix` option for all our operators defaulting to their/one Kubernetes short name would be good. I think I'd like that.

**Comment by maltesander** [comment]
So i did some investigation. Currently the only thing we use via name (and not labels) is the create_config_map method. There we check if the config map already exists / data differs etc. Querying pods already happens via labels (e.g. spark operator to check if workers have all current masters set in their command).

I would suggest a combined approach of your ideas. I like having it at least a little bit readable. 
What do you think of ("[x-y]" would be the amount of characters reserved):
<short-name[1-5]>-<cluster-name[1-15]>-<role[1-10]>-(<data/config[1-10]>)-<kubernetes-generated-name>
That would e.g. look like:
Pod: zk-production-server-xxxxxxxx
Cm: zk-production-server-data-xxxxxxx

We could still add the remaining parts of the node name, but i think normally that would be too cryptic anyways.
This is readable and we can fall back to the generated name from kubernetes.

What do you think?

**Comment by lfrancke** [comment]
That sounds sensible.

My original idea was to generate the "full" name and then truncate at the end but I like yours better because it retains more information.
This feels like a method for operator-rs?

So this'd mean we need a couple of follow-up tickets if we decide to go this way:
- Implement a generic function to generate the name (in operator-rs)
- Change all operators to use that function
- Change all operators to fix create_config_map to not use names (we need some way to identify the correct CM then....)

Am I missing something?

**Comment by maltesander** [comment]
Yes i think that is all. We wanted to move the create_config_maps in the operator-rs anyways (I think @soenkeliebau did that already with the command refactoring). So we would change it there and adapt the operators.

I can create the tickets and would start to adapt / create the get_pod_name method to our specifications.

**Comment by lfrancke** [comment]
Thank you! Go ahead.

**Comment by maltesander** [comment]
TODO for operators:
- [ ] OPA
- [ ] Spark
- [ ] ZooKeeper
- [ ] Kafka
- [ ] Monitoring
- [ ] NiFi

**Comment by maltesander** [comment]
Implemented in operator-rs and all operators.
